### PR TITLE
Add HTTP/2 Server Push Support

### PR DESCRIPTION
**Pull request by thuva9872**
## Purpose
Add HTTP/2 server push feature to the existing HTTP WebSocket inbound endpoint.

## Goals
Adding HTTP/2 server push support.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/synapse/config/xml/XMLConfigConstants.java [hunk 1]
```diff
@@ -135,4 +135,8 @@ public static enum DATA_TYPES {
     public static final String XPATH_ENVELOPE_RELATIVE = "envelope";
     
     public static final String CONFIG_REF = "configKey";
+    /**
+     * The server push processing sequence name
```
Comment (Nirothipan) [inline modules/core/src/main/java/org/apache/synapse/config/xml/XMLConfigConstants.java:139]:
```suggestion
     * The server push processing sequence name.
```

Comment (Nirothipan) [inline modules/core/src/main/java/org/apache/synapse/config/xml/XMLConfigConstants.java:139]:
```suggestion
     * The server push processing sequence name.
```

File: modules/core/src/main/java/org/apache/synapse/config/xml/endpoints/HTTPEndpointFactory.java [hunk 2]
```diff
@@ -143,6 +144,18 @@ protected Endpoint createEndpoint(OMElement epConfig, boolean anonymousEndpoint,
...
+            OMAttribute versionAttr = httpElement.getAttribute(new QName("version"));
+            if (versionAttr != null) {
+                setHttpVersion(httpEndpoint, versionAttr.getAttributeValue());
+            }
+            if (BridgeConstants.HTTP_2_0_VERSION.equals(httpEndpoint.getHttpVersion())) {
+                OMElement serverPushSequence =
+                        httpElement.getFirstChildWithName(new QName(SynapseConstants.SYNAPSE_NAMESPACE,
+                                XMLConfigConstants.SERVER_PUSH_SEQUENCE));
+                if (serverPushSequence != null && serverPushSequence.getText() != null) {
+                    httpEndpoint.setServerPushSequence(serverPushSequence.getText());
```
Comment (Nirothipan) [inline modules/core/src/main/java/org/apache/synapse/config/xml/endpoints/HTTPEndpointFactory.java:156]: serverPushSequence.getText().trim()

File: modules/core/src/main/java/org/apache/synapse/endpoints/HTTPEndpoint.java [hunk 3]
```diff
@@ -38,15 +40,22 @@
 
 public class HTTPEndpoint extends AbstractEndpoint {
 
+    private final String DEFAULT_HTTP_VERSION = "1.1";
```
Comment (Nirothipan) [inline modules/core/src/main/java/org/apache/synapse/endpoints/HTTPEndpoint.java:43]: can use HTTP_1_1_VERSION here ?

File: modules/core/src/main/java/org/apache/synapse/core/axis2/Axis2FlexibleMEPClient.java [hunk 4]
```diff
@@ -443,6 +444,12 @@ public static void send(
             if (endpoint.isUseSeparateListener()) {
                 axisOutMsgCtx.getOptions().setUseSeparateListener(true);
             }
+
+            //Set server push sequence in axis message context if available
```
Comment (Nirothipan) [inline modules/core/src/main/java/org/apache/synapse/core/axis2/Axis2FlexibleMEPClient.java:448]:
```suggestion
            //Set server push sequence in axis message context if available.
```

File: modules/core/src/main/java/org/apache/synapse/core/axis2/SynapseCallbackReceiver.java [hunk 5]
```diff
@@ -778,4 +789,36 @@ private void handleNoCallback(String messageID, MessageContext messageCtx){
...
+        synCtx.setProperty(BridgeConstants.IS_PUSH_PROMISE, messageCtx.getProperty(BridgeConstants.IS_PUSH_PROMISE));
+        synCtx.setProperty(BridgeConstants.SERVER_PUSH_RESOURCE_PATH,
+                messageCtx.getProperty(BridgeConstants.SERVER_PUSH_RESOURCE_PATH));
+        TenantInfoConfigurator configurator = synCtx.getEnvironment().getTenantInfoConfigurator();
+        if (configurator != null) {
+            configurator.extractTenantInfo(synCtx);
+        }
+        SequenceMediator serverPushSequence =
+                (SequenceMediator) messageCtx.getProperty(BridgeConstants.SERVER_PUSH_SEQUENCE);
+        if (serverPushSequence != null && synCtx != null) {
```
Comment (Nirothipan) [inline modules/core/src/main/java/org/apache/synapse/core/axis2/SynapseCallbackReceiver.java:820]: synCtx != null -- this should already be true, else we will get NPE before.

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/Axis2ServerPushListener.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import org.wso2.transport.http.netty.message.Http2PushPromise;
+import org.wso2.transport.http.netty.message.HttpCarbonMessage;
+
+public class Axis2ServerPushListener implements HttpClientConnectorListener {
+
+    private static final Log LOG = LogFactory.getLog(Axis2ServerPushListener.class);
+
+    private HttpResponseFuture future;
+    private HttpClientConnector clientConnector;
+    private MessageContext requestMsgCtx;
```
Comment (Nirothipan) [inline modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/Axis2ServerPushListener.java:38]: these variables can be final ?

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/Axis2ServerPushListener.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    private HttpCarbonMessage clientRequest;
+
+    public Axis2ServerPushListener(HttpResponseFuture future, HttpClientConnector clientConnector,
+                                   MessageContext requestMsgCtx, WorkerPool workerPool) {
+
+        this.future = future;
+        this.clientConnector = clientConnector;
+        this.requestMsgCtx = requestMsgCtx;
+        this.workerPool = workerPool;
+        clientRequest =
```
Comment (Nirothipan) [inline modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/Axis2ServerPushListener.java:50]: we're not using this variable ?

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/HttpServerPushWorker.java [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * {@code HttpServerPushWorker} is the Thread which does the server push processing.
+ */
+public class HttpServerPushWorker implements Runnable {
+
+    private static final Log LOG = LogFactory.getLog(HttpServerPushWorker.class);
+    private Http2PushPromise pushPromise;
```
Comment (Nirothipan) [inline modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/HttpServerPushWorker.java:49]: can be final ?

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/HttpServerPushWorker.java [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+    private HttpCarbonMessage httpResponse;
+    private MessageContext requestMsgCtx;
```
Comment (Nirothipan) [inline modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/HttpServerPushWorker.java:51]: can be final ?

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/SourceResponseHandler.java [hunk 10]
```diff
@@ -419,6 +436,34 @@ public static void sendResponse(MessageContext msgCtx, HttpCarbonMessage inbound
...
     }
 
+    /**
+     * Send the server push promise to the client
+     *
+     * @param http2PushPromise Http2PushPromise
+     * @param clientRequest    HttpCarbonMessage
+     * @throws AxisFault if error occurred while pushing responses.
+     */
+    public static void pushPromise(Http2PushPromise http2PushPromise, HttpCarbonMessage clientRequest) throws AxisFault {
```
Comment (Nirothipan) [inline modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/SourceResponseHandler.java:446]: Method comment should end with a period, the variables and returned types need to be explained. shall we check and fix other places as well.

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/Axis2ServerPushListener.java (hunk 7 above)
Comment (thuva9872) [reply modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/netty/sender/Axis2ServerPushListener.java:50]: removed

**Review by Nirothipan — APPROVED** [review]

**Pull request merged**